### PR TITLE
Dataviews: Add preview and grid view in templates list

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Part of: https://github.com/WordPress/gutenberg/issues/55083

This PR add the `grid` view support to templates list. By default(in `list` view) the field is hidden. I'm not sure how we can make it more performant, since the previews are expensive and block the UI. Aside of this PR, we need a better solution to handle the previews and I think folks are exploring some approaches.

## Testing Instructions
1. Enable the admin views experiment and go to manage all templates view
2. Play around with the `grid` view and in `list` view make the field visible and do the same

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/16275880/e914eaf3-0a2e-4291-8707-e7ecc528bc2f

### Notes
I've added some `TODOs` in the code that could be done in followups. I think that's fine for now, that the feature is experimental and is heavily iterated.

### CODE REVIEW

**Comment by jameskoster** [comment]
This is pretty neat to see.

I think there's a fundamental question to answer about whether the previews should be included in list view since their usefulness is diminished at smaller sizes. Nevertheless I've pushed a couple of small design adjustments:

* Reduced preview size.
* Positioned to the left of the Title for visual balance.
* Added a fixed height and border/radius.

Screenshot 2023-11-21 at 15 33 47

**Comment by ntsekouras** [comment]
Thank you for polishing @jameskoster !

>I think there's a fundamental question to answer about whether the previews should be included in list view since their usefulness is diminished at smaller sizes. 

I think that's something we can handle separately. In general the main take away here, is that we will probably need a way to have different fields per view type(we don't have it right now). This is probably going to be useful for mobile.

**Comment by jameskoster** [comment]
I pushed some changes to the preview dimensions in grid view. It's tricky to get the dimensions right accounting for different 'length' templates (note the visible white space at the bottom of the 'Product search results template' in the screenshot below) but this feels like a reasonable starting point from which to refine. 

Screenshot 2023-11-22 at 15 20 19

@ntsekouras I don't know if it's feasible, but if we could apply the theme background color to `.block-editor-block-preview__container` I think that would help.

**Comment by ntsekouras** [comment]
>@ntsekouras I don't know if it's feasible, but if we could apply the theme background color to .block-editor-block-preview__container I think that would help.

I think we have a good starting point and we could refine in [this PR](https://github.com/WordPress/gutenberg/pull/56441).

File: packages/edit-site/src/components/dataviews/style.scss [hunk 1]
```diff
@@ -38,6 +38,19 @@
 	tr {
 		border-bottom: 1px solid $gray-100;
 	}
+	.block-editor-block-preview__container {
```
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/style.scss]: This style seem misplaced to me, the data views is generic and this is for a specific component.

File: packages/edit-site/src/components/dataviews/view-actions.js [hunk 2]
```diff
@@ -39,6 +39,8 @@ const availableViews = [
 		id: 'grid',
 		label: __( 'Grid' ),
 	},
+	// TODO: This(the whole layout) needs be redesigned because in order for it to
+	// work properly, needs extra code from the consumer of DataViews component.
```
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/view-actions.js]: What is this comment about?

File: packages/edit-site/src/components/dataviews/style.scss [hunk 3]
```diff
@@ -38,6 +38,19 @@
...
+		border-radius: $radius-block-ui;
+	}
+}
+
+.dataviews-grid-view {
+	.block-editor-block-preview__container {
+		height: 320px;
+		border: 1px solid $gray-300;
+		border-radius: $radius-block-ui;
+	}
```
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/style.scss]: conditionally styling a field should be implemented in the field, not the view. That's the reason the "render" function of a field receives the "view" object as well.

File: packages/edit-site/src/components/dataviews/view-grid.js [hunk 4]
```diff
@@ -23,12 +24,23 @@ export function ViewGrid( { data, fields, view, actions, getItemId } ) {
...
+		>
+			{ shownData.map( ( item, index ) => {
 				return (
 					<VStack key={ getItemId?.( item ) || index }>
 						<div className="dataviews-view-grid__media">
+							{ /* TODO: This needs to be handled better because it could be not possible to
+								to return `null`, if the field needs to provide a component that uses hooks, etc..
+								In that case, the actual field could render nothing, but the `mediaField?.render`
+								call would return a React element.
+							*/ }
```
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/view-grid.js]: Good catch, we should definitely fix that quickly.

File: packages/edit-site/src/components/dataviews/view-grid.js [hunk 5]
```diff
@@ -50,7 +62,11 @@ export function ViewGrid( { data, fields, view, actions, getItemId } ) {
 									) ) }
 								</VStack>
 							</FlexBlock>
-							<FlexBlock>
+							{ /* TODO: ItemActions needs to be handled better in general.
+							In smaller viewports the actions could take too much space.
+							A solution could be to render the actions inside the media,
+							or in `grid` render all actions in the drop down menu. */ }
```
Comment (youknowriad) [inline packages/edit-site/src/components/dataviews/view-grid.js]: Should we make these todo lists as tasks in the issue instead of leaving them in the code

File: packages/edit-site/src/components/page-templates/dataviews-templates.js [hunk 6]
```diff
@@ -199,27 +237,46 @@ export default function DataviewsTemplates() {
...
-				actions={ actions }
-				data={ shownTemplates }
-				getItemId={ ( item ) => item.id }
-				isLoading={ isLoadingData }
-				view={ view }
-				onChangeView={ onChangeView }
-				supportedLayouts={ [ 'list' ] }
-			/>
-		</Page>
+		<ExperimentalBlockEditorProvider settings={ settings }>
```
Comment (youknowriad) [inline packages/edit-site/src/components/page-templates/dataviews-templates.js]: Can we start by moving this to the field. I know it will be duplicated but if it has no performance impact (to check), I think it's more understandable.

File: packages/edit-site/src/components/dataviews/style.scss (hunk 1 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/style.scss]: You're totally right! I'll move them.

File: packages/edit-site/src/components/dataviews/view-actions.js (hunk 2 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/view-actions.js]: Right now, the view actions menu shows the available layouts from here, including the `side by side`. But since this layout needs the consumer to add code, it makes no sense to show it by default, as it does nothing without the extra code.

Makes sense?

File: packages/edit-site/src/components/dataviews/view-grid.js (hunk 5 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/dataviews/view-grid.js]: I'll do that. In general I was planning to work on follow ups right after this.

File: packages/edit-site/src/components/dataviews/view-actions.js (hunk 2 above)
Comment (youknowriad) [reply packages/edit-site/src/components/dataviews/view-actions.js]: I don't know, for me it's fine, its design is not ready yet (it needs to be a list) which IMO will remove the confusion.

**Comment by oandregal** [comment]
As it is now, the grid layout I felt it was a bit too confusing (perhaps a side effect of my testing theme being based on a white/black palette):

https://github.com/WordPress/gutenberg/assets/583546/4518ddf3-bf89-4035-babb-14b0ed903795

So I merged [56441](https://github.com/WordPress/gutenberg/pull/56441) into this one to test whether those changes would help in separating visually the cards. The end result looks like this:

https://github.com/WordPress/gutenberg/assets/583546/2d064b8f-7eca-4a48-ac64-14745a51478f

I just thought I'd bring it up, no sure what the action steps would be. I think we should have a grid for templates, though more visual order/hierarchy (specially for white/black based themes) would be helpful.

File: packages/edit-site/src/components/page-templates/dataviews-templates.js [hunk 7]
```diff
@@ -31,17 +36,28 @@ import {
...
+
 const DEFAULT_VIEW = {
 	type: 'list',
 	search: '',
 	page: 1,
 	perPage: 20,
 	// All fields are visible by default, so it's
 	// better to keep track of the hidden ones.
-	hiddenFields: [],
+	hiddenFields: [ 'preview' ],
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates/dataviews-templates.js:60]: I still consider a bug that the preview field cannot be hidden in the grid layout :/ Not to fix in this PR, though.

https://github.com/WordPress/gutenberg/assets/583546/9d9b797c-9e9e-4f3c-8e73-fc8ac6c573bc

File: packages/edit-site/src/components/page-templates/dataviews-templates.js (hunk 6 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-templates/dataviews-templates.js]: I moved it and there doesn't seem to be an impact - this doesn't mean that the previews performance is good though :)

File: packages/edit-site/src/components/page-templates/dataviews-templates.js [hunk 8]
```diff
@@ -94,6 +110,32 @@ function AuthorField( { item } ) {
 	);
 }
 
+function TemplatePreview( { content, viewType } ) {
+	const settings = usePatternSettings();
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates/dataviews-templates.js:114]: It's a bit weird that we use this hook when this page is unrelated to patterns. Looking at the data the hook provides, it sounds like we'd only need the settings from the site editor (everything else is related to patterns). Can we pull those settings directly here?

**Review by youknowriad — COMMENTED** [review]
This is looking good to me code wise.

File: packages/edit-site/src/components/page-templates/dataviews-templates.js [hunk 9]
```diff
@@ -153,6 +195,22 @@ export default function DataviewsTemplates() {
 	}, [ allTemplates, view ] );
 	const fields = useMemo(
 		() => [
+			{
+				header: __( 'Preview' ),
+				id: 'preview',
+				getValue: () => {},
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates/dataviews-templates.js]: Do we need this?

File: packages/edit-site/src/components/page-templates/dataviews-templates.js [hunk 10]
```diff
@@ -153,6 +195,22 @@ export default function DataviewsTemplates() {
...
+				render: ( { item, view: { type: viewType } } ) => {
+					return (
+						<TemplatePreview
+							content={ item.content.raw }
+							viewType={ viewType }
+						/>
+					);
+				},
+				minWidth: 120,
```
Comment (oandregal) [inline packages/edit-site/src/components/page-templates/dataviews-templates.js:209]: I mentioned elsewhere, but this data is only for one of the layouts (list). We should find a different way to provide this. Either by making it explicit in the field API (e.g.: scope it under `layout.list.minWidth`, for example) or by absorbing it in the view config (in principle, I'd prefer this option, as it relates to controlling layout).

Not to fix in this PR, but important to figure out before stabilizing/creating a package for dataviews.

**Comment by jameskoster** [comment]
@oandregal I think some styles are missing there, the template previews should have a static height to avoid breaking the grid and 'longer' templates causing excessive scrolling.

File: packages/edit-site/src/components/page-templates/dataviews-templates.js (hunk 10 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-templates/dataviews-templates.js:209]: Agreed, but let's explore this separately. Probably it should be enough by just adding specific css on each field, per view type.

File: packages/edit-site/src/components/page-templates/dataviews-templates.js (hunk 8 above)
Comment (youknowriad) [reply packages/edit-site/src/components/page-templates/dataviews-templates.js:114]: This hook is misnamed I think. It's about providing the block editor with the settings it needs to render anything (with all the patterns, styles...)

Comment (ntsekouras) [reply packages/edit-site/src/components/page-templates/dataviews-templates.js:114]: Actually this hook loads the patterns are needed here too, in order to render previews with patterns inside. These patterns are loaded within a Block Editor instance and in our case we don't load one.

File: packages/edit-site/src/components/page-templates/dataviews-templates.js (hunk 9 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/page-templates/dataviews-templates.js]: It seems we don't :). I removed it.

**Comment by ntsekouras** [comment]
Is something more needed to land this? 🤔

**Comment by youknowriad** [comment]
Looks good to me. 

 - you definitely feel the lagginess more. static previews are going to solve this I think.
 - I think the icon for the source field could be hidden in Grid view maybe.

**Review by youknowriad — APPROVED** [review]

**Comment by ntsekouras** [comment]
>I think the icon for the source field could be hidden in Grid view maybe.

You mean in the view type menu on top? Or for the author? Probably the latter, right?

**Comment by youknowriad** [comment]
for the author, sorry.

**Comment by jameskoster** [comment]
It might be good to simplify the icon more generally (no background, 24px footprint, smaller gap between icon + label):

Screenshot 2023-11-23 at 15 25 14

Then it might work a bit better in grid view. Hiding is still an option though.

**Comment by ntsekouras** [comment]
If we hide it we'll have the issue we were discussing about ambiguity. I'll try the smaller icon and we can discuss in that new PR.
